Transfer command

jcarpanelli opened the pull request:
Fixes #823 

I've started implementing a command in the cli to transfer an amount of ether/wei/gwei between accounts.
For this, I've created a `TransactionController` in the `cli` package that will probably host almost all functions that will interact with the `Transactions` lib's object.

A demo of how the command works:

![transfer](https://i.ibb.co/BTdjMb2/transfer.gif)

spalladino commented on packages/cli/src/scripts/transfer.ts:
```diff
@@ -0,0 +1,8 @@
+import ScriptError from '../models/errors/ScriptError';
+import { CreateParams } from './interfaces';
+
+export default async function createProxy({ units, to, value, txParams = {} }: any): Promise<any> {
```
```suggestion
export default async function transfer({ units, to, value, txParams = {} }: any): Promise<any> {
```

spalladino commented on packages/cli/src/models/network/TransactionController.ts:
```diff
@@ -0,0 +1,18 @@
+    log.info(`Sending ${value} wei to ${to}...`);
+    const receipt = await Transactions.sendRawTransaction(to, { value }, this.txParams);
+    log.info(`Funds successfully sent!. Receipt:`);
+    stdout(receipt);
```
Let's output just the tx hash instead of the full receipt

spalladino commented on packages/cli/src/commands/transfer.ts:
```diff
@@ -0,0 +1,61 @@
+  .command(signature, undefined, { noHelp: true })
+  .usage('[options]')
+  .description(description)
+  .option('--units [units]', 'specify units.')
```
```suggestion
  .option('--units <units>', 'specify units.')
```

I'd also consider renaming to `unit` instead of `units`, and indicate which values are valid ('eth', 'wei', etc).

spalladino commented on packages/cli/src/commands/transfer.ts:
```diff
@@ -0,0 +1,61 @@
+        value: account
+      }))
+    },
+    units: {
```
WDYT about removing this question and just defaulting to ETH? The user can still use a different unit if using the option explicitly.

jcarpanelli replied:
I agree. This way we can also let the user choose between any valid unit.

spalladino commented on packages/cli/src/commands/transfer.ts:
```diff
@@ -0,0 +1,61 @@
+  const { txParams } = await ConfigVariablesInitializer.initNetworkConfiguration(promptedNetwork, true);
+  const accounts = await ZWeb3.accounts();
+  const opts = { units, to, from, value };
+  console.log(opts);
```
```suggestion
```

jcarpanelli commented:
Ready for review!
I refactored and finished implementing the command. I also removed the `unit` type prompt question. Now, if the user do not specify any unit type, `ether` will be the default, and if the user specifies an non-existent unit, an error will be thrown. Note that we're allowing the user to specify any unit that is [here](https://web3js.readthedocs.io/en/1.0/web3-utils.html#id74). 
Updated demo:

![image](https://user-images.githubusercontent.com/3752824/56931913-1d52c080-6ab8-11e9-829a-1be3c301537b.png)

spalladino commented on packages/cli/src/commands/transfer.ts:
```diff
@@ -0,0 +1,63 @@
+  .description(description)
+  .option('--to <to>', 'specify recipient address')
+  .option('--value <value>', 'the amount of ether units to be transferred')
+  .option('--unit <unit>', 'unit name. If none is given, \'ether\' will be used. To see supported units, go to: https://web3js.readthedocs.io/en/1.0/web3-utils.html#id85')
```
Let's avoid tying ourselves to web3 documentation. I'd just list here the most common ones (wei, kwei, gwei, milli, ether).

spalladino commented on packages/cli/src/commands/transfer.ts:
```diff
@@ -0,0 +1,63 @@
+    },
+    value: {
+      type: 'input',
+      message: 'Enter an amount to transfer',
```
Can we include the `unit` as part of the message, so we remind the user right here?

jcarpanelli replied:
We will be reminding them the unit while they type the amount value (with the `transformer` function)

spalladino commented on packages/cli/src/models/network/TransactionController.ts:
```diff
@@ -0,0 +1,23 @@
+
+  public async transfer(to: string, amount: string, unit: string): Promise<void> {
+    if (!isValidUnit(unit)) {
+      throw Error(`Invalid specified unit ${unit}. To see supported units, visit: https://web3js.readthedocs.io/en/1.0/web3-utils.html#id74`);
```
Same here, let's just list the most common ones

spalladino commented on packages/cli/test/commands/share.js:
```diff
@@ -85,33 +87,7 @@ exports.stubCommands = function () {
-    this.getZosversion.restore()
-    this.packageFile.restore()
-    this.dependency.restore()
+    sinon.restore()
```
:clap:

ylv-io replied:
The moment you feel stupid!

spalladino commented on packages/cli/test/scripts/transfer.test.js:
```diff
@@ -0,0 +1,51 @@
+
+    context('when specifying an invalid unit', function() {
+      it('throws an error', async function() {
+        await transfer({ to: receiver, value: '10', unit: 'palla' }).should.be.rejectedWith(/Invalid specified unit palla/);
```
:smile:

spalladino commented on packages/cli/src/prompts/prompt.ts:
```diff
@@ -24,6 +24,7 @@ interface InquirerQuestion {
   default?: any;
   choices?: ChoicesT;
   when?: ((answers: { [key: string]: any }) => boolean);
+  transformer?: ((value: string, answers: { [key: string]: any }) => string);
```
What's the difference between this one and normalize?

jcarpanelli replied:
`normalize` is a custom function defined by us to specify the format of the answer, while `transformer` is an `inquirer.js` function that  returns a transformed value to display in the console. The latter only impacts what is shown while editing and it doesn't modify the answer. You can check the gif above, and see how the `ether`, `gwei`, etc is shown while typing an amount value.

ylv-io commented on packages/cli/test/scripts/transfer.test.js:
```diff
@@ -0,0 +1,51 @@
+
+        await transfer({ from: sender, to: receiver, value: '10', unit: 'ether' });
+
+        (await ZWeb3.getBalance(sender)).should.not.eq(110e18.toString());
```
I assume you are not using `90e18` because of gas cost? I think we should either calculate gas cost or use epsilon if it is not possible, like `|90e18 - value| < epsilon`. WDYT?

jcarpanelli replied:
Indeed, that's the reason!  I also noticed I'm asserting against `110e18` instead of `100e18` because of your review! I think we are fine leaving it as `100e18` for now.

spalladino commented:
> Why `ZWeb3.getBalance` returns `string` when `eth.getBalance` returns `BigNumber`?

Note that, in web3 beta 37, getBalance returns a string and not a BN. However, I agree that we should move into BN - the question is which one. Given the last contributions by george on ethers, I think it's time to revisit that.
